Python/Product/Analysis/Interpreter/Ast/AstAnalysisFunctionWalker.cs

```diff
@@ -111,25 +102,139 @@ FunctionDefinition targetFunction
+                    if (value is AstPythonConstant c && c.Type is AstPythonTuple tuple) {
+                        var types = tuple.Types.ToArray();
+                        var names = tex.Items.Select(x => (x as NameExpression)?.Name).ToArray();
+                        for (var i = 0; i < Math.Min(names.Length, types.Length); i++) {
+                            if (names[i] != null) {
+                                _scope.SetInScope(names[i], new AstPythonConstant(types[i]));
+                            }
+                        }
+                    }
+                }
+            }
+            return base.Walk(node);
+        }
+
+        public override bool Walk(IfStatement node) {
+            // Handle basic check such as
+            // if isinstance(value, type):
+            //    return value
+            // by assigning type to the value unless clause is raising exception.
+            var ce = node.Tests.FirstOrDefault()?.Test as CallExpression;
+            if (ce?.Target is NameExpression ne && ne?.Name == "isinstance" && ce.Args.Count == 2) {
+                var name = (ce.Args[0].Expression as NameExpression)?.Name;
+                var typeName = (ce.Args[1].Expression as NameExpression)?.Name;
+                if (name != null && typeName != null) {
+                    var typeId = BuiltinTypeIdFromName(typeName);
+                    if (typeId != BuiltinTypeId.Unknown) {
+                        _scope.SetInScope(name, new AstPythonConstant(new AstPythonBuiltinType(typeName, typeId)));
+                    }
+                }
+            }
             return base.Walk(node);
         }
 
         public override bool Walk(ReturnStatement node) {
-            foreach (var type in _scope.GetTypesFromValue(_scope.GetValueFromExpression(node.Expression))) {
-                _returnTypes.Add(type);
+            var typesAdded = false;
+            if (node.Expression is NameExpression nex && nex.Name == "self") {
+                // For self return the actual class without added private members
+                var klass = _scope.LookupNameInScopes("__class__", NameLookupContext.LookupOptions.Nonlocal);
+                if (klass != null) {
+                    _returnTypes.Add(klass as IPythonType);
+                    typesAdded = true;
+                }
+            }
+
+            if (!typesAdded) {
+                foreach (var type in _scope.GetTypesFromValue(_scope.GetValueFromExpression(node.Expression))) {
+                    _returnTypes.Add(type);
+                }
+            }
+            return true; // We want to evaluate all code so all private variables in __new__ get defined
+        }
+
+        private IMember GetClass() {
+            bool classmethod, staticmethod;
+            GetMethodType(_target, out classmethod, out staticmethod);
+            var klass = _scope.LookupNameInScopes("__class__", NameLookupContext.LookupOptions.Local);
```

Comment:
Because of the `.Local`, this will only be useful in the constructor before we do the first `PushScope()`. Perhaps we should label it as such? Or make it use `NonLocal` and call it later in the constructor

Reply:
This code is a fragment extracted as is from the original `public void Walk()`.